File: exp-15/q3-lps-table/solution.c
#include <stdio.h>
#include <string.h>

void trim(char *s)
{
    s[strcspn(s, "\r\n")] = '\0';
}

int main()
{
    char pattern[1000];
    fgets(pattern, sizeof(pattern), stdin);
    trim(pattern);
    int m = strlen(pattern);
    int lps[1000];
    lps[0] = 0;
    int len = 0, i = 1;
    while (i < m)
    {
        if (pattern[i] == pattern[len])
        {
            len++;
            lps[i] = len;
            i++;
        }
        else if (len != 0)
        {
            len = lps[len - 1];
        }
        else
        {
            lps[i] = 0;
            i++;
        }
    }
    printf("LPS:");
    for (i = 0; i < m; i++)
        printf(" %d", lps[i]);
    printf("\n");
    return 0;
}

File: exp-15/q3-lps-table/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	pattern := scanner.Text()
	m := len(pattern)
	lps := make([]int, m)
	length, i := 0, 1
	for i < m {
		if pattern[i] == pattern[length] {
			length++
			lps[i] = length
			i++
		} else if length != 0 {
			length = lps[length-1]
		} else {
			lps[i] = 0
			i++
		}
	}
	fmt.Print("LPS:")
	for i = 0; i < m; i++ {
		fmt.Printf(" %d", lps[i])
	}
	fmt.Println()
}
